extensions/standard-processors/processors/TailFile.cpp

```diff
@@ -204,52 +418,78 @@ void TailFile::parseStateFileLine(char *buf) {
   }
 
   std::string value = equal;
-  key = trimRight(key);
-  value = trimRight(value);
+  key = utils::StringUtils::trimRight(key);
+  value = utils::StringUtils::trimRight(value);
 
   if (key == "FILENAME") {
     std::string fileLocation, fileName;
     if (utils::file::PathUtils::getFileNameAndPath(value, fileLocation, fileName)) {
       logger_->log_debug("State migration received path %s, file %s", fileLocation, fileName);
-      tail_states_.insert(std::make_pair(fileName, TailState { fileLocation, fileName, 0, 0 }));
+      state.emplace(fileName, TailState{fileLocation, fileName});
     } else {
-      tail_states_.insert(std::make_pair(value, TailState { fileLocation, value, 0, 0 }));
+      state.emplace(value, TailState{fileLocation, value});
     }
   }
   if (key == "POSITION") {
     // for backwards compatibility
-    if (tail_states_.size() != 1) {
+    if (tail_states_.size() != std::size_t{1}) {
       throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, "Incompatible state file types");
     }
     const auto position = std::stoull(value);
     logger_->log_debug("Received position %d", position);
-    tail_states_.begin()->second.currentTailFilePosition_ = position;
+    state.begin()->second.position_ = position;
   }
   if (key.find(CURRENT_STR) == 0) {
     const auto file = key.substr(strlen(CURRENT_STR));
     std::string fileLocation, fileName;
     if (utils::file::PathUtils::getFileNameAndPath(value, fileLocation, fileName)) {
-      tail_states_[file].path_ = fileLocation;
-      tail_states_[file].current_file_name_ = fileName;
+      state[file].path_ = fileLocation;
+      state[file].file_name_ = fileName;
     } else {
       throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, "State file contains an invalid file name");
     }
   }
 
   if (key.find(POSITION_STR) == 0) {
     const auto file = key.substr(strlen(POSITION_STR));
-    tail_states_[file].currentTailFilePosition_ = std::stoull(value);
+    state[file].position_ = std::stoull(value);
   }
 }
 
+bool TailFile::recoverState(const std::shared_ptr<core::ProcessContext>& context) {
+  std::map<std::string, TailState> new_tail_states;
+  bool state_load_success = getStateFromStateManager(new_tail_states) ||
+                            getStateFromLegacyStateFile(context, new_tail_states);
+  if (!state_load_success) {
+    return false;
+  }
 
+  logger_->log_debug("load state succeeded");
 
-bool TailFile::recoverState(const std::shared_ptr<core::ProcessContext>& context) {
-  bool state_load_success = false;
+  if (tail_mode_ == Mode::SINGLE) {
+    if (tail_states_.size() == 1) {
+      auto state_it = tail_states_.begin();
+      const auto it = new_tail_states.find(state_it->first);
+      if (it != new_tail_states.end()) {
+        state_it->second = it->second;
+      }
+    } else {
+      throw minifi::Exception(ExceptionType::PROCESSOR_EXCEPTION, "This should never happen: "
```

Comment:
I don't think it can happen, as a configuration change will restart the processor.  A newly created processor in Single file mode starts with a one-element state, and we only update that one item here, we don't add to it.  This `throw` is here only to guard against later code changes which change this behavior.

As to invalidating the state, I'll need to look at how NiFi does that.  I think there is value in keeping the state of existing files after a configuration change: for example, if the selection is broadened in Multi file mode (say from `*.log` to `*.*`), we probably don't want to restart the files we have already tailed from the beginning.  But we should do whatever NiFi does.  Can that be a separate pull request?

Reply:
Yep, it's fine to do as a follow-up.